### DISCUSSION TITLE
Trying to run a LAMMPS simulation with two deepmd potentials

### DISCUSSION THREAD

**Discussion by Constant206462** [deepmodeling/deepmd-kit#4371] [Q&A]
Hi. I'm on deepmd v2.1.5. I encountered a problem, while trying to simulate black phosphorene sheets on top of quartz in LAMMPS using deepmd potentials.
There are two potentials I want to use. The first one was trained to predict interactions between all three atom types (Si, O, P). The second one predicts only the behavior of phosphorene.
An important note here: the datasets with which the models are trained are not the same, they are collected from different ab initio MD simulations.
The models' type maps are as follows:
Si 0
O 1
P 2
and
P 0

For LAMMPS simulation, I set types to be
Si 1
O 2
P 3

So what I want is to use the SiOP potential for interaction between types 1\*2 and 1\*2\*3 (Si-Si, Si-O, Si-P, O-O, O-P), and P potential for 3 3 (P-P) interaction.

_The first question: is it even possible?_ If not, the further message is meaningless.

Although I managed to launch a simulation, I don't know if it is correct. What worked for me now is this
`pair_style	hybrid/overlay deepmd sio2p.pb deepmd phosp.pb`
`pair_coeff	* * deepmd 1`
`pair_coeff	* * deepmd 2 NULL NULL P`

The log file shows this
pair_coeff	* * deepmd 2 NULL NULL P
WARNING: Interaction between types 0 and 2 is set with deepmd, but will be ignored.
 Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/pair_deepmd.cpp:1042)
WARNING: Interaction between types 0 and 3 is set with deepmd, but will be ignored.
 Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/pair_deepmd.cpp:1042)
WARNING: Interaction between types 1 and 2 is set with deepmd, but will be ignored.
 Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/pair_deepmd.cpp:1042)
WARNING: Interaction between types 1 and 3 is set with deepmd, but will be ignored.
 Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/pair_deepmd.cpp:1042)
WARNING: Interaction between types 2 and 2 is set with deepmd, but will be ignored.
 Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/pair_deepmd.cpp:1042)
WARNING: Interaction between types 2 and 3 is set with deepmd, but will be ignored.
 Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/pair_deepmd.cpp:1042)
WARNING: Interaction between types 3 and 3 is set with deepmd, but will be ignored.
 Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/libdeepmd_1663923207577/work/source/lmp/pair_deepmd.cpp:1042)

_The second question: does the above mean, that the P potential is actually computes forces for Si-Si (type 1) interaction?_
In this case, I should somehow remap the types in the SiOP model, shouldn't I?

_The third question: if it works properly (the P potential computes P-P interaction), how do I remove the SiOP potential's contribution in P-P interaction?_
Ideally, I would use `pair_style hybrid` instead of `pair_style hybrid/overlay`, but when I set `hybrid`, I get an error:
ERROR: Pair hybrid sub-style is not used (src/pair_hybrid.cpp:572)
Which is just weird. As if the P potential never computes anything in the first place.

**Comment by njzjz** [comment]
> What worked for me now is this
> `pair_style hybrid/overlay deepmd sio2p.pb deepmd phosp.pb`
> `pair_coeff * * deepmd 1`
> `pair_coeff * * deepmd 2 NULL NULL P`

Assigning the type map via pair_coeff (https://github.com/deepmodeling/deepmd-kit/pull/2267) was added in v2.2.0. `NULL` support (#2732) is added in v2.2.4. I don't believe those will work in v2.1.5.

**Reply by Constant206462** [reply]
I just installed deepmd v2.2.11 and converted my models via dp convert-from. After trying to run the same script, this error poped out
ERROR: pair_coeff: element P not found in the model (/home/conda/feedstock_root/build_artifacts/deepmd-kit_1720339766254/work/source/lmp/pair_deepmd.cpp:1258)
Last command: pair_coeff	* * deepmd 2 NULL NULL P
Does that mean I have to retrain the models with the newer deepmd version? Or is there a way to add mapping to the model without retraining?
Thanks!

**Reply by njzjz** [reply]
Which elements are in your model?

**Reply by Constant206462** [reply]
In sio2p.pb it's Si O and P. In phosp.pb it's only P.

**Reply by njzjz** [reply]
Could you print the type map?

```py
from deepmd import (
    DeepPotential,
)
dp = DeepPotential("phosp.pb")
print("type_map:", dp.get_type_map())
```

**Reply by Constant206462** [reply]
Sure. This is the output
`type_map: []`
Which is understandable considering the error message from earlier.

**Reply by Constant206462** [reply]
I have another question. I guess it's not possible to change the type_map in a model, but is it possible to copy a network weights from one model into the other if the models' architecture is identical, except that one has a different type_map?
The model I want to copy the weights from is a frozen pb model. Unfortunately, the training checkpoints for this one were deleted a long time ago.

**Reply by njzjz** [reply]
> is it possible to copy a network weights from one model into the other if the models' architecture is identical

The `--init-frz-model ` option in `dp train` can do it.

**Reply by njzjz** [reply]
Alternatively, you can use `dp convert-from` to convert the `.pb` file to a `.pbtxt` file, modify it, and then convert it back to `.pb`. It might be not easy to get started, though.

**Reply by Constant206462** [reply]
> The `--init-frz-model ` option in `dp train` can do it.

Thanks! I think that worked.

**Reply by Constant206462** [reply]
Off to initial questions. I tried to run my MD script with `pair_style hybrid` and different `pair_coef` options:
`pair_coeff	* * deepmd 2 NULL NULL P`
`pair_coeff	* * deepmd 2 P`
`pair_coeff	* * deepmd 2`
None seem to work. The output is pretty much the same for each launch, except for the first one on the list, for which the warnings don't show up. The error message is the same.
> <...>
>   \>\>\> Info of model(s):
>   using   1 model(s): ../../phosp.pb 
>   rcut in model:      6
>   ntypes in model:    1
> WARNING: Interaction between types 1 and 2 is set with deepmd, but will be ignored.
>  Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/deepmd-kit_1720339766254/work/source/lmp/pair_deepmd.cpp:1282)
> WARNING: Interaction between types 1 and 3 is set with deepmd, but will be ignored.
>  Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/deepmd-kit_1720339766254/work/source/lmp/pair_deepmd.cpp:1282)
> WARNING: Interaction between types 2 and 2 is set with deepmd, but will be ignored.
>  Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/deepmd-kit_1720339766254/work/source/lmp/pair_deepmd.cpp:1282)
> WARNING: Interaction between types 2 and 3 is set with deepmd, but will be ignored.
>  Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/deepmd-kit_1720339766254/work/source/lmp/pair_deepmd.cpp:1282)
> WARNING: Interaction between types 3 and 3 is set with deepmd, but will be ignored.
>  Deepmd model has only 1 types, it only computes the mulitbody interaction of types: 1-1. (/home/conda/feedstock_root/build_artifacts/deepmd-kit_1720339766254/work/source/lmp/pair_deepmd.cpp:1282)
> <...>
> ERROR: Pair hybrid sub-style deepmd is not used (src/pair_hybrid.cpp:599)
> Last command: minimize	0.0 1.0e-8 1000 100000

The script still launches with `pair_style hybrid/overlay`. Although the warnings are here as well, and I wonder if phosp.pb even contributes to interaction.
Is there a way to run it with `pair_style hybrid` or somehow exclude the sio2p.pb model contribution to P-P interaction?
